src/domain/login/LoginViewModel.ts

```diff
@@ -15,34 +15,51 @@ limitations under the License.
 */
 
 import {Client} from "../../matrix/Client.js";
-import {ViewModel} from "../ViewModel";
+import {Options as BaseOptions, ViewModel} from "../ViewModel";
 import {PasswordLoginViewModel} from "./PasswordLoginViewModel.js";
 import {StartSSOLoginViewModel} from "./StartSSOLoginViewModel.js";
 import {CompleteSSOLoginViewModel} from "./CompleteSSOLoginViewModel.js";
 import {LoadStatus} from "../../matrix/Client.js";
 import {SessionLoadViewModel} from "../SessionLoadViewModel.js";
 
-export class LoginViewModel extends ViewModel {
-    constructor(options) {
+// TODO(isaiah): make these default exported
+import {PasswordLoginMethod} from "../../matrix/login/PasswordLoginMethod";
+import {SSOLoginHelper} from "../../matrix/login/SSOLoginHelper";
+import {TokenLoginMethod} from "../../matrix/login/TokenLoginMethod";
+
+type Options = {
+    defaultHomeserver: string;
+    ready: ReadyFn;
+    loginToken?: string;
+} & BaseOptions;
+
+export class LoginViewModel extends ViewModel<Options> {
+    private _ready: ReadyFn;
+    private _loginToken?: string;
+    private _client: Client;
+    private _loginOptions?: LoginOptions;
+    private _passwordLoginViewModel?: PasswordLoginViewModel;
+    private _startSSOLoginViewModel?: StartSSOLoginViewModel;
+    private _completeSSOLoginViewModel?: CompleteSSOLoginViewModel;
+    private _loadViewModel?: SessionLoadViewModel;
+    private _loadViewModelSubscription?: () => void;
+    private _homeserver: string;
+    private _queriedHomeserver?: string;
+    private _abortHomeserverQueryTimeout?: () => void;
+    private _abortQueryOperation?: () => void;
+
+    private _hideHomeserver = false;
+    private _isBusy = false;
+    private _errorMessage = "";
+
+    constructor(options: Readonly<Options>) {
         super(options);
         const {ready, defaultHomeserver, loginToken} = options;
         this._ready = ready;
         this._loginToken = loginToken;
         this._client = new Client(this.platform);
-        this._loginOptions = null;
-        this._passwordLoginViewModel = null;
-        this._startSSOLoginViewModel = null;
-        this._completeSSOLoginViewModel = null;
-        this._loadViewModel = null;
-        this._loadViewModelSubscription = null;
         this._homeserver = defaultHomeserver;
-        this._queriedHomeserver = null;
-        this._errorMessage = "";
-        this._hideHomeserver = false;
-        this._isBusy = false;
-        this._abortHomeserverQueryTimeout = null;
-        this._abortQueryOperation = null;
-        this._initViewModels();
+        void this._initViewModels();
```

Comment:
OOI, why do you do this? Because it's an async method?

Reply:
Yes, I've been using the `.ts-eslintrc.js` which was complaining at me
```
Promises must be handled appropriately or explicitly marked as ignored with the `void` operator.eslint[@typescript-eslint/no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/v4.29.2/packages/eslint-plugin/docs/rules/no-floating-promises.md)
```

Normally I would mark the function `async` and `await` it, but it's a constructor so I just `void`ed it.